8359477: com/sun/net/httpserver/Test12.java appears to have a temp file race

jaikiran opened the pull request:
Can I please get a review of this test-only change which addresses an intermittent failure in `test/jdk/com/sun/net/httpserver/Test12.java`? This fixes https://bugs.openjdk.org/browse/JDK-8359477.

In the linked JBS issue I've added a comment which explains what causes this intermittent failure. The issue is a bug/race within the test. There are 2 commits in this PR and only of them is necessary for addressing this issue. The other one is a general cleanup of the test. I am willing to revert the general cleanup commit from this test if others feel that it shouldn't be done in this PR.

The actual fix is this commit https://github.com/openjdk/jdk/commit/a0f8fc806e1682ef909cb7b4a449be855072fe48 which stops deleting the input files upon test completion. That should be OK because the jtreg test infrastructure will clean them up from the scratch directory after the completion of the test run (if the test succeeds). There are other ways to address this race condition and continue to delete these files (for example, co-ordinating between the test and the test specific `FileServerHandler`), but to me it didn't seem worthwhile doing that.

I've run this change around a 1000 times and haven't seen it fail.

P.S: The `test/jdk/com/sun/net/httpserver/Test1.java` does the same things as this test, but does it sequentially. I'm pretty sure that we should be able to trigger this intermittent test failure in that test too if I reorder the `test()` calls in that test to have a fixed length test be the last one to execute (i.e. `test (true, ...)` instead of the current `test(false, ...)`).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8359477](https://bugs.openjdk.org/browse/JDK-8359477): com/sun/net/httpserver/Test12.java appears to have a temp file race (**Bug** - P4)


### Reviewers
 * [Volkan Yazici](https://openjdk.org/census#vyazici) (@vy - Committer) Review applies to [42b86dee](https://git.openjdk.org/jdk/pull/25820/files/42b86deebba9aee237904d0b4391306ed2416a38)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25820/head:pull/25820` \
`$ git checkout pull/25820`

Update a local copy of the PR: \
`$ git checkout pull/25820` \
`$ git pull https://git.openjdk.org/jdk.git pull/25820/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25820`

View PR using the GUI difftool: \
`$ git pr show -t 25820`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25820.diff">https://git.openjdk.org/jdk/pull/25820.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25820#issuecomment-2975401009)
</details>

vy commented on test/jdk/com/sun/net/httpserver/Test12.java:
```diff
@@ -99,95 +94,76 @@ public static void main (String[] args) throws Exception {
             if (s2 != null)
                 s2.stop(0);
             if (executor != null)
-                executor.shutdown ();
-            Files.delete(smallFilePath);
-            Files.delete(largeFilePath);
```
Instead of removing the file cleanup, IMHO, _good practice_, can't we use `shutdown() + awaitTermination()` or `shutdownNow()` for the executor?

jaikiran replied:
Hello Volkan, the issue here's isn't about not waiting for the tasks to complete. They in fact do complete. The actual issue is that the tasks have no way to know if the server side has actually done a `FileInputStream.close()` for the fixed length case, because the `InputStream` handed out to those tasks doesn't wait to receive a EOF from the server side for the fixed length case. Instead since the InputStream knows how many bytes to expect, once it receives those many bytes, it immediately returns. The server side (in FileServerHandler) would have transmitted those bytes and would be in the process of closing the FileInputStream, but the client side task would already be completed. So adding a `awaitTermination()` on the executor will not help.

jaikiran replied:
> Instead of removing the file cleanup, IMHO, good practice

They do get cleaned up - jtreg cleans up the scratch directories (unless asked not to do so) when the test completes. The general pattern we use in JDK tests is to let jtreg clean them up for us so that if the test has failed (for whatever reason), then due to the use of `-retain:fail,error`, these input files are retained by jtreg so that they can help debug any unexpected test failures. 
In this case, the input files themselves very likely won't play a role in any unexpected failures. So if we still prefer to delete these explicitly in the test, then I can fix this in a different way -  that will require a bit more code for handshake between the test's `FileServerHandler` and the test itself to ensure that the test knows when it is OK to delete these files.

vy replied:
> the issue here's isn't about not waiting for the tasks to complete. They in fact do complete. The actual issue is that the tasks have no way to know if the server side has actually done a `FileInputStream.close()` for the fixed length case

@jaikiran, `executor` is configured for the servers via `s1.setExecutor (executor)` and `s2.setExecutor (executor)` lines above. I was under the impression that shutting down the executor (i.e., interrupting the server handlers) would result in server releasing all acquired resources, including issuing a `FIS::close`. Was I mistaken?

vy replied:
Thanks, now I see. Those resource usages in `FileServerHandler` should ideally be wrapped in try-with-resources, but that is out of the scope of this work.

jaikiran replied:
> Each new request will be submitted to the executor by the HttpServer, therefore waiting for the executor to finish should ensure that FileServerHandler::handle has finished, and therefore that the the file has been closed.

That's a good point (both your and Volkan's). I missed this detail when Volkan mentioned it because I focused on the part which noted the server would close the file. I will experiment with this (or a different variant that I had in mind) since it looks like the preference is to let this delete calls stay.

jaikiran commented:
The PR is now ready for review. When I opened this PR a few weeks back I had removed the calls to `Files.delete(...)` in order to not introduce bigger changes to the test to co-ordinate the file deletion. However, Volkan's suggested approach of calling `ExecutorService.close()` to ensure that the server side handlers complete execution before deleting the files is a clean and straightforward way to address this issue. So i have re-introduced the Files.delete(...) calls and replaced the use of `ExecutorService.shutdown()` with `ExecutorService.close()` in the test. That's the real fix in this PR, the rest of the changes are clean ups to use newer constructs like try-with-resources.

With the current changes in this PR, I have run the test around a 1000 times and haven't seen it fail. Plus, I've also run the test with an aritifcial slowdown in the `FileServerHandler` and the test itself and I could reproduce the issue consistently without the proposed changes. It no longer reproduces even with artificial delays after the proposed changes.

dfuch commented on test/jdk/com/sun/net/httpserver/Test12.java:
```diff
@@ -49,27 +38,32 @@
         Path smallFilePath = createTempFileOfSize(TEMP_FILE_PREFIX, null, 23);
         Path largeFilePath = createTempFileOfSize(TEMP_FILE_PREFIX, null, 2730088);
-        try {
+        try (final ExecutorService executor = Executors.newCachedThreadPool()) {
```
I would not use try-with-resources here: I don't think we should close the executor before closing the servers.

jaikiran replied:
I've updated the PR to close the Executor after the servers have been closed. I'm guessing the suggestion was mostly as a precaution than what this test currently does?

msheppar commented:
the ubiquitous copyright update for FileServerHandler.java ;-)

jaikiran commented:
Thank you everyone for the inputs and reviews.

jaikiran commented:
/integrate
